### ISSUE TITLE
Uploading a large script causes a crash (bad stack size)

### ISSUE THREAD

**Issue by CrashOverride85** [CrashOverride85/zc95#91]
See https://github.com/CrashOverride85/zc95/discussions/81#discussioncomment-9132948

Notes mostly for future me.

Uploading the script attached to the above discussions causes core0 to crash (hardfault).

## Observations:
* Disabling stack guards (PICO_USE_STACK_GUARDS) causes the upload to work
* On crashing, the stack pointer isn't close to limit, so it doesn't appear to be a stack overflow
* PICO_STACK_SIZE is set to 12288. _Reducing_ it to 8192 causes the upload to work, but breaks remotely running some scripts due to stack overflow (stack pointer at limit).

Although it doesn't appear to be explictly documented in the pico C SDK, the limit for PICO_STACK_SIZE seems to be 8k (multicore complicates things, so the limit is sort-of 4kb).
Going beyond that means the end of the stack - and the write protected section of memory that is the stack guard - ends up in memory that can be used for the heap. 

### Cause of crash
Due to the stack/heap overlap, malloc is returning memory which includes the stack guards, and as soon as that gets written to => hardfault. So not a stack overflow, but the heap expanding to the area assigned to the stack.

## Fix
Options:
### 1. Expand stack properly
Keep `PICO_STACK_SIZE=12288`, and create a custom linker script which is just a copy/paste of the SDK `memmap_default.ld` named `memmap_zc95.ld`, with:
`RAM(rwx) : ORIGIN =  0x20000000, LENGTH = 256k ` 
changed to:
`RAM(rwx) : ORIGIN =  0x20000000, LENGTH = 244k `

Then added to `CMakeLists.txt` as:
`pico_set_linker_script(zc95 ${CMAKE_SOURCE_DIR}/memmap_zc95.ld)`

This means the heap can't ever expand into the region where the stack (and stack guard) is, effectively working around the 8kb stack limit.

Trying to upload the script with this change results in the error:
```
Opening: /dev/ttyUSB1
Connection opened
Uploading: |██████████████████████████████████████████████████| 100.0% Complete

Got error message: 
    Invalid script: not enough memory
Connection closed
Failed
```
So significant improvement - it's no longer crashing, and is giving a reasonable error. And mostly confirms that it is the heap growing into the stack as the cause, as the 4kb reduction is enough to cause malloc to fail (note that `PICO_MALLOC_PANIC=0`, hence Lua being able to handle this rather than a panic).

### 2. Reduce stack usage
Set PICO_STACK_SIZE to 8192, and fix the issue where a lot of memory is being (unnecessarily!) used on the stack when generating messages to send of serial/wifi in remote access mode.

Going with 2 as it actually fixes the issue, but it's good to understand _why_ it was going so wrong.

**Comment by CrashOverride85** [comment]
Mostly fixed in branch fix_large_script_crash, but need to have a rethink about how large scripts are uploaded. ATM the full 48k is allocated in memory at the start of the upload, then written to flash at the end.
It probably want's changing to write it in blocks, with a marker somewhere to indicate it's a valid script (maybe set the first block to nulls or something until it's finished).

**Comment by CrashOverride85** [comment]
Fixed in dev, fix will be in version >=1.9
